docs/decisions/0002-ledger-balance-enforcement.rst

```diff
@@ -0,0 +1,346 @@
+    - **Accepted** *(date)* once it is agreed upon
+    - **Superseded** *(date)* with a reference to its replacement if a later ADR changes or reverses the decision
+
+    If an ADR has Draft status and the PR is under review, you can either use the intended final status (e.g. Provisional, Accepted, etc.), or you can clarify both the current and intended status using something like the following: "Draft (=> Provisional)". Either of these options is especially useful if the merged status is not intended to be Accepted.
+
+Context
+*******
+
+The Ledger model enables the bookkeeping of all value ins/outs for a given subsidy; value can be added and removed from
+a Ledger via the creation of positive-, or negative-value Transactions.  The "balance" of a Ledger is simply the sum of
+all of its Transactions.  One critical requirement for the Ledger is that it not be allowed to enter a state of negative
+balance, which is enforced via the following code pattern:
+
+.. code-block:: python
+
+  with atomic(durable=True):
+      if (quantity < 0) and ((ledger.balance() + quantity) < 0):
+          # throw error
+      # create a transaction with the given quantity
+
+However, without leveraging any database or other guardrails to prevent race conditions, the sample code alone does not
+guarantee an always-positive balance.  For example, two requests may attempt a subsidy redemption close in time:
+
++-----------------+-----------------------+-----------------------+
+| Ledger Balance  |  Request 1            | Request 2             |
++=================+=======================+=======================+
+|             10  |                       |                       |
++-----------------+-----------------------+-----------------------+
+|                 | check balance         |                       |
++-----------------+-----------------------+-----------------------+
+|                 |                       | check balance         |
++-----------------+-----------------------+-----------------------+
+|                 | create -7 transaction |                       |
++-----------------+-----------------------+-----------------------+
+|              3  |                                               |
++-----------------+-----------------------+-----------------------+
+|                 |                       | create -4 transaction |
++-----------------+-----------------------+-----------------------+
+|             -1  |                       |                       |
++-----------------+-----------------------+-----------------------+
+
+Guardrails provided by Django and MySQL by default are insufficient to prevent this scenario.  Django may implicitly
+wrap the request in a MySQL database transaction (or we may explicitly do so via `atomic()` context manager), but a
+MySQL transaction will at most implicitly grab a row-level lock.  Simple row-level locks, however, are insufficient in
```

Comment:
request: if you know of one or two offhand, got a link to docs about this?

Reply:
added links.